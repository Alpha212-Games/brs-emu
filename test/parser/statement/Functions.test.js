const Parser = require("../../../lib/parser");
const { Lexeme } = require("../../../lib/Lexeme");
const BrsError = require("../../../lib/Error");

const { EOF } = require("../ParserTests");

describe("parser", () => {
    afterEach(() => BrsError.reset());

    describe("function declarations", () => {
        it("parses minimal empty function declarations", () => {
            let parsed = Parser.parse([
                { kind: Lexeme.Function, text: "function", line: 1 },
                { kind: Lexeme.Identifier, text: "foo", line: 1 },
                { kind: Lexeme.LeftParen, text: "(", line: 1 },
                { kind: Lexeme.RightParen, text: ")", line: 1 },
                { kind: Lexeme.Newline, text: "\\n", line: 1 },
                { kind: Lexeme.EndFunction, text: "end function", line: 2 },
                EOF
            ]);

            expect(BrsError.found()).toBeFalsy();
            expect(parsed).toBeDefined();
            expect(parsed).not.toBeNull();
            expect(parsed).toMatchSnapshot();
        });
    });

    describe("sub declarations", () => {
        it("parses minimal sub declarations", () => {
            let parsed = Parser.parse([
                { kind: Lexeme.Sub, text: "sub", line: 1 },
                { kind: Lexeme.Identifier, text: "bar", line: 1 },
                { kind: Lexeme.LeftParen, text: "(", line: 1 },
                { kind: Lexeme.RightParen, text: ")", line: 1 },
                { kind: Lexeme.Newline, text: "\\n", line: 1 },
                { kind: Lexeme.EndSub, text: "end sub", line: 2 },
                EOF
            ]);

            expect(BrsError.found()).toBeFalsy();
            expect(parsed).toBeDefined();
            expect(parsed).not.toBeNull();
            expect(parsed).toMatchSnapshot();
        });
    });
});
