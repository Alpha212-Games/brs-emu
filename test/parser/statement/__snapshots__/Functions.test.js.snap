// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parser function declarations parses functions with default argument expressions 1`] = `
Array [
  Function {
    "body": Block {
      "statements": Array [],
    },
    "name": Object {
      "kind": 21,
      "line": 1,
      "text": "add",
    },
    "parameters": Array [
      Object {
        "defaultValue": Literal {
          "value": Int32 {
            "kind": 3,
            "value": 3,
          },
        },
        "name": "a",
        "type": 8,
      },
      Object {
        "defaultValue": Literal {
          "value": Int32 {
            "kind": 3,
            "value": 4,
          },
        },
        "name": "b",
        "type": 8,
      },
      Object {
        "defaultValue": Binary {
          "left": Variable {
            "name": Object {
              "kind": 21,
              "line": 1,
              "text": "a",
            },
          },
          "right": Literal {
            "value": Int32 {
              "kind": 3,
              "value": 5,
            },
          },
          "token": Object {
            "kind": 8,
            "line": 1,
            "text": "+",
          },
        },
        "name": "c",
        "type": 8,
      },
    ],
  },
]
`;

exports[`parser function declarations parses functions with implicit-dynamic arguments 1`] = `
Array [
  Function {
    "body": Block {
      "statements": Array [],
    },
    "name": Object {
      "kind": 21,
      "line": 1,
      "text": "add2",
    },
    "parameters": Array [
      Object {
        "defaultValue": undefined,
        "name": "a",
        "type": 8,
      },
      Object {
        "defaultValue": undefined,
        "name": "b",
        "type": 8,
      },
    ],
  },
]
`;

exports[`parser function declarations parses functions with typed arguments 1`] = `
Array [
  Function {
    "body": Block {
      "statements": Array [],
    },
    "name": Object {
      "kind": 21,
      "line": 1,
      "text": "repeat",
    },
    "parameters": Array [
      Object {
        "defaultValue": undefined,
        "name": "str",
        "type": 2,
      },
      Object {
        "defaultValue": undefined,
        "name": "count",
        "type": 3,
      },
    ],
  },
]
`;

exports[`parser function declarations parses minimal empty function declarations 1`] = `
Array [
  Function {
    "body": Block {
      "statements": Array [],
    },
    "name": Object {
      "kind": 21,
      "line": 1,
      "text": "foo",
    },
    "parameters": Array [],
  },
]
`;

exports[`parser sub declarations parses minimal sub declarations 1`] = `
Array [
  Function {
    "body": Block {
      "statements": Array [],
    },
    "name": Object {
      "kind": 21,
      "line": 1,
      "text": "bar",
    },
    "parameters": Array [],
  },
]
`;

exports[`parser sub declarations parses subs with default argument expressions 1`] = `
Array [
  Function {
    "body": Block {
      "statements": Array [],
    },
    "name": Object {
      "kind": 21,
      "line": 1,
      "text": "add",
    },
    "parameters": Array [
      Object {
        "defaultValue": Literal {
          "value": Int32 {
            "kind": 3,
            "value": 3,
          },
        },
        "name": "a",
        "type": 8,
      },
      Object {
        "defaultValue": Literal {
          "value": Int32 {
            "kind": 3,
            "value": 4,
          },
        },
        "name": "b",
        "type": 8,
      },
      Object {
        "defaultValue": Binary {
          "left": Variable {
            "name": Object {
              "kind": 21,
              "line": 1,
              "text": "a",
            },
          },
          "right": Literal {
            "value": Int32 {
              "kind": 3,
              "value": 5,
            },
          },
          "token": Object {
            "kind": 8,
            "line": 1,
            "text": "+",
          },
        },
        "name": "c",
        "type": 8,
      },
    ],
  },
]
`;

exports[`parser sub declarations parses subs with implicit-dynamic arguments 1`] = `
Array [
  Function {
    "body": Block {
      "statements": Array [],
    },
    "name": Object {
      "kind": 21,
      "line": 1,
      "text": "add2",
    },
    "parameters": Array [
      Object {
        "defaultValue": undefined,
        "name": "a",
        "type": 8,
      },
      Object {
        "defaultValue": undefined,
        "name": "b",
        "type": 8,
      },
    ],
  },
]
`;

exports[`parser sub declarations parses subs with typed arguments 1`] = `
Array [
  Function {
    "body": Block {
      "statements": Array [],
    },
    "name": Object {
      "kind": 21,
      "line": 1,
      "text": "repeat",
    },
    "parameters": Array [
      Object {
        "defaultValue": undefined,
        "name": "str",
        "type": 2,
      },
      Object {
        "defaultValue": undefined,
        "name": "count",
        "type": 3,
      },
    ],
  },
]
`;
